src/Qualify.hs

```diff
@@ -291,104 +320,93 @@ qualifyLst typeEnv globalEnv env (XObj (Lst xobjs) i t) =
 qualifyLst _ _ _ xobj = Left $ FailedToQualifySymbols xobj
 
 -- | Qualify a single symbol.
--- TODO: Clean this up
 qualifySym :: Qualifier
-qualifySym typeEnv globalEnv localEnv xobj@(XObj (Sym path _) i t) =
-  Right $
-    Qualified $
-      case path of
-        -- Unqualified:
-        SymPath [] name ->
-          case lookupBinder path (getTypeEnv typeEnv) of
-            Just (Binder _ (XObj (Lst (XObj (Interface _ _) _ _ : _)) _ _)) ->
-              -- Found an interface with the same path!
-              -- Have to ensure it's not a local variable with the same name as the interface
-              case lookupInEnv path localEnv of
-                Just (foundEnv, _) ->
-                  if envIsExternal foundEnv
-                    then createInterfaceSym name
-                    else doesNotBelongToAnInterface False localEnv
-                Nothing ->
-                  --trace ("Will turn '" ++ show path ++ "' " ++ prettyInfoFromXObj xobj ++ " into an interface symbol.")
-                  createInterfaceSym name
-            _ ->
-              doesNotBelongToAnInterface False localEnv
-        -- Qualified:
-        _ ->
-          doesNotBelongToAnInterface False localEnv
+-- Unqualified path.
+qualifySym typeEnv globalEnv localEnv xobj@(XObj (Sym path@(SymPath _ name) _) i t) =
+  ( ( ( replaceLeft
```

Comment:
It's similar to the old version of this function, but I tried to make it slightly cleaner by performing lookups first (what happens in the first part), then based on the result, resolving/qualifying the symbol appropriately.

There's a lot going on here and it can probably be improved even further/made more clear in future commits. 

Here's the gist of what the entire function does:

1. Try the following lookups on the unqualified symbol, taking the first one that succeeds:
    1. Look for the symbol directly in the global type environment.
    2. Look for the symbol directly in the local environment (usually a module or function env)
    3. Look for the symbol directly in the global environment
    4. Look for the symbol in all the `use`d environments of the local environment
    5. Look for the symbol "contextually" (current env + all parents)
    6. Finally, if all else turned up nothing, look for the symbol in every possible env, starting from the global env (lookupEverywhere)
2. Based on which lookup succeeds, we'll either get a single binder and resolve that `resolve`, or we'll get multiple binders (in a list), (i.e. there's more than one possible match) and resolve those `resolveMulti`
    1. `resolve` just performs special checks so that we qualify the binder in the right way in case there's some edge cases:
        1. Check if the binder is an interface, if so, ensure it's not shadowed by some other local binding. If it isn't, return an `InterfaceSym`
        2. Check if the binder is an external override binding
        3. Check if the binding is the name of a module (assume it's naked init of a type)
        4. For all other binders, check the env mode of where the binder is defined and whether or not the binding is captured, and assign the lookup mode appropriately.
    2. `resolveMulti` Attempts to add all the candidate paths to a `MultiSym`, but first it ensures:
        1. The input path wasn't already qualified (since if this is true, we should know precisely what symbol we need)
        2. There are no local definitions among the candidates (none of the possible path matches are defined in an environment with mode local) since this would mean an environment "leaked" into a broader scope.

Reply:
Great, I'll refer back to this if/when I have to touch this code :)